File: bp-agent/src-go/tools/registry.go
package tools

import (
	"fmt"
	"sync"
)

// ToolRegistry manages tool registration and execution.
type ToolRegistry struct {
	mu    sync.RWMutex
	tools map[string]ToolEntry
}

// NewToolRegistry creates a new ToolRegistry.
func NewToolRegistry() *ToolRegistry {
	return &ToolRegistry{tools: map[string]ToolEntry{}}
}

// Register adds a new tool to the registry.
func (r *ToolRegistry) Register(name string, handler ToolHandler, schema ToolSchema) error {
	if name == "" || handler == nil {
		return fmt.Errorf("invalid tool")
	}
	if schema.Name == "" {
		schema.Name = name
	}
	if schema.Name != name {
		return fmt.Errorf("schema name mismatch")
	}

	r.mu.Lock()
	defer r.mu.Unlock()
	if _, ok := r.tools[name]; ok {
		return fmt.Errorf("tool already registered")
	}
	r.tools[name] = ToolEntry{Name: name, Handler: handler, Schema: schema}
	return nil
}

// Execute runs a tool by name with args.
func (r *ToolRegistry) Execute(name string, args map[string]any) ToolResult {
	r.mu.RLock()
	entry, ok := r.tools[name]
	r.mu.RUnlock()
	if !ok {
		return ToolResult{Success: false, Error: "tool not found"}
	}
	out, err := entry.Handler(args)
	if err != nil {
		return ToolResult{Success: false, Error: err.Error()}
	}
	return ToolResult{Success: true, Output: out}
}

// GetSchemas returns the registered tool schemas.
func (r *ToolRegistry) GetSchemas() []ToolSchema {
	r.mu.RLock()
	defer r.mu.RUnlock()
	schemas := make([]ToolSchema, 0, len(r.tools))
	for _, entry := range r.tools {
		schemas = append(schemas, entry.Schema)
	}
	return schemas
}

// Has returns true if a tool is registered.
func (r *ToolRegistry) Has(name string) bool {
	r.mu.RLock()
	defer r.mu.RUnlock()
	_, ok := r.tools[name]
	return ok
}

// Count returns the number of registered tools.
func (r *ToolRegistry) Count() int {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return len(r.tools)
}

File: bp-agent/src-rs/tools/registry.rs
use std::collections::HashMap;
use std::sync::RwLock;

use serde_json::Value;

use super::types::{ToolEntry, ToolHandler, ToolResult, ToolSchema};

pub struct ToolRegistry {
    tools: RwLock<HashMap<String, ToolEntry>>,
}

impl ToolRegistry {
    pub fn new() -> Self {
        Self {
            tools: RwLock::new(HashMap::new()),
        }
    }

    pub fn register(&self, name: &str, handler: ToolHandler, mut schema: ToolSchema) -> Result<(), String> {
        if name.is_empty() {
            return Err("invalid tool".to_string());
        }
        if schema.name.is_empty() {
            schema.name = name.to_string();
        }
        if schema.name != name {
            return Err("schema name mismatch".to_string());
        }

        let mut map = self.tools.write().map_err(|_| "lock error".to_string())?;
        if map.contains_key(name) {
            return Err("tool already registered".to_string());
        }
        map.insert(
            name.to_string(),
            ToolEntry {
                name: name.to_string(),
                handler,
                schema,
            },
        );
        Ok(())
    }

    pub fn execute(&self, name: &str, args: Value) -> ToolResult {
        let map = match self.tools.read() {
            Ok(lock) => lock,
            Err(_) => {
                return ToolResult {
                    success: false,
                    output: None,
                    error: Some("lock error".to_string()),
                }
            }
        };

        let entry = match map.get(name) {
            Some(entry) => entry,
            None => {
                return ToolResult {
                    success: false,
                    output: None,
                    error: Some("tool not found".to_string()),
                }
            }
        };

        match (entry.handler)(args) {
            Ok(output) => ToolResult {
                success: true,
                output: Some(output),
                error: None,
            },
            Err(err) => ToolResult {
                success: false,
                output: None,
                error: Some(err),
            },
        }
    }

    pub fn get_schemas(&self) -> Vec<ToolSchema> {
        let map = match self.tools.read() {
            Ok(lock) => lock,
            Err(_) => return vec![],
        };
        map.values().map(|entry| entry.schema.clone()).collect()
    }

    pub fn has(&self, name: &str) -> bool {
        let map = match self.tools.read() {
            Ok(lock) => lock,
            Err(_) => return false,
        };
        map.contains_key(name)
    }

    pub fn count(&self) -> usize {
        let map = match self.tools.read() {
            Ok(lock) => lock,
            Err(_) => return 0,
        };
        map.len()
    }
}
